File: frontend/src/types/conversation.ts
/**
 * Conversation domain types.
 *
 * Matches API_SPEC §34-41 (Conversations), §44-48 (Group Members).
 * Spec reference: MASTER_PROJECT_SPEC §14, §15 (Conversation + Membership)
 */

import type { UserRef } from "./user";
import type { Message } from "./message";

/** Re-exported for consumer convenience — conversation components use UserRef for member avatars. */
export type { UserRef };


export type ConversationType = "direct" | "group";
export type MemberRole = "admin" | "member";

export interface ConversationMember {
  user_id: string;
  display_name: string;
  avatar_url: string | null;
  role: MemberRole;
  joined_at: string; // ISO 8601
}

export interface Conversation {
  id: string;
  type: ConversationType;
  name: string | null;
  avatar_url: string | null;
  last_message: Pick<Message, "id" | "content" | "sender_id" | "created_at"> | null;
  unread_count: number;
  updated_at: string; // ISO 8601
  members?: ConversationMember[];
}

export interface ConversationDetail extends Conversation {
  created_by: string;
  created_at: string;
  members: ConversationMember[];
}

export interface CreateDirectRequest {
  user_id: string;
}

export interface CreateGroupRequest {
  name: string;
  member_ids: string[];
}

export interface ConversationListResponse {
  data: Conversation[];
  meta: { count: number };
}

export interface ConversationResponse {
  data: ConversationDetail;
}

File: backend/app/schemas/conversation.py
from typing import List, Optional
from datetime import datetime
from pydantic import BaseModel, ConfigDict
from .user import UserResponse

class MessagePreview(BaseModel):
    id: str
    content: str
    sender_id: str
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)

class ConversationListItem(BaseModel):
    id: str
    type: str
    name: Optional[str] = None
    avatar_url: Optional[str] = None
    last_message: Optional[MessagePreview] = None
    unread_count: int = 0
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)

class ConversationMemberSchema(BaseModel):
    id: str
    display_name: str
    role: str
    model_config = ConfigDict(from_attributes=True)

class ConversationDetail(BaseModel):
    id: str
    type: str
    name: Optional[str] = None
    avatar_url: Optional[str] = None
    created_by: str
    members: List[ConversationMemberSchema] = []
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)

class CreateDirectRequest(BaseModel):
    user_id: str

class CreateGroupRequest(BaseModel):
    name: str
    member_ids: List[str]

class AddMemberRequest(BaseModel):
    user_id: str
